contracts/AgentMech.sol

```diff
@@ -96,19 +117,31 @@ contract AgentMech is ERC721Mech {
 
         // Increase the number of undelivered requests
         numUndeliveredRequests++;
+        // Increase the total number of requests
+        numTotalRequests++;
 
         emit Request(msg.sender, requestId, data);
     }
 
+    /// @dev Performs actions before the delivery of a request.
+    /// @param data Self-descriptive opaque data-blob.
+    /// @return requestData Data for the request processing.
+    function _preDeliver(uint256, bytes memory data) internal virtual returns (bytes memory requestData) {
+        requestData = data;
+    }
+
     /// @dev Delivers a request.
     /// @param requestId Request id.
     /// @param data Self-descriptive opaque data-blob.
     function deliver(uint256 requestId, bytes memory data) external onlyOperator {
+        // Perform a pre-delivery of the data if it needs additional parsing
+        bytes memory requestData = _preDeliver(requestId, data);
+
         // Remove delivered request Id from the request Ids map
         uint256[2] memory requestIds = mapRequestIds[requestId];
-        // Check if the request Id is invalid: previous and next request Ids are zero,
-        // and the zero's element next request Id is not equal to the provided request Id
-        if (requestIds[0] == 0 && requestIds[1] == 0 && mapRequestIds[0][1] != requestId) {
+        // Check if the request Id is invalid (non existent or delivered): previous and next request Ids are zero,
+        // and the zero's element previous request Id is not equal to the provided request Id
+        if (requestIds[0] == 0 && requestIds[1] == 0 && mapRequestIds[0][0] != requestId) {
```

Comment:
Why did this change?

Reply:
If this condition is true, mapRequestIds[0][0] and mapRequestIds[0][1] are the same. Just decided to have the simplest layout of the array.